### DISCUSSION TITLE
Read Eddystone TLM advertisement data

### DISCUSSION THREAD

**Discussion by Ripper346** [hbldh/bleak#1310] [Q&A] [answered]
Hi, after struggling I was able to make the scanner working and now I wanted to grab the advertisement of some eddystone beacons. I thought that the returning AdvertisementData object when something is discovered included it but it doesn't seem so. For example this is an advertisement I received
```
AdvertisementData(service_data={'0000feaa-0000-1000-8000-00805f9b34fb': b' \x00\x0b\xa1\x14\x90\x00\xa0FV\x0f\xe3Ut'}, service_uuids=['0000feaa-0000-1000-8000-00805f9b34fb'], rssi=-76)
```
I am 100% sure that the beacon is sending eddystone tlm packets but the only thing I am seeing is that one service data that it is too short to be it. Is there a way to see the data (even raw) received? Thank you

**Comment by dlech** [comment]
What other data were you expecting? I see the service UUID and service data that is described at https://github.com/google/eddystone/blob/master/protocol-specification.md and the service data is 14 bytes as described at https://github.com/google/eddystone/blob/master/eddystone-tlm/tlm-plain.md.

**Reply by Ripper346** [reply]
Sorry, I am failing to see the 10 bytes of the common element, even if it is supposed to be the service uuid key. How can you say that it is an eddystone tlm? The only thing I recognize is the `feaa`. I don't know if I have to convert the uuid key in some way. In other words, for that packet shouldn't I expect something like this? `\x02\x01\x06\x03\x03\xaa\xfe\x11\x16\xaa\xfe \x00\x0b\xa1\x14\x90\x00\xa0FV\x0f\xe3Ut`

**Reply by dlech** [reply]
Something like this?

```python
...
import struct
...

TLM_UUID = '0000feaa-0000-1000-8000-00805f9b34fb'

...

def handle_adv_data(device, adv_data):
    tlm = adv_data.service_data[TLM_UUID]
    frame_type, version, voltage, temperature, count, time = struct.unpack("<BBHHII", tlm)
    ...
```

**Reply by Ripper346** [reply]
Oh, so you are just assuming that that uuid corresponds to an eddystone tlm package. Ok, however the unpacking is wrong, using beacontools the example package says voltage 2977 and temperature 205625, but the unpacking says 41227 and temperature 32788 (I don't have 33°C in the room, it would be quite unpleasant 🥵). I think there is something wrong but I have never used the struct package, I need to check the documentation to understand the string you wrote

**Reply by dlech** [reply]
Also check https://github.com/google/eddystone/blob/master/eddystone-tlm/tlm-plain.md#field-notes. It looks like it uses some weird encodings that I didn't take into account. You will probably have to manipulate the unpacked values since I don't think there are any struct formats that correspond to some of the stuff used.

**Reply by Ripper346** [reply]
OK, thank you very much!

**Reply by Ripper346** [reply, marked as answer]
Ok, I looked into beacontools library and the code below gives the correct values. I write it so if anyone land on the question has a complete answer.
```python
from construct import Struct, Int16ub, Int32ub, GreedyRange, Byte

UnencryptedTLMFrame = Struct(
    "frame_type" / Byte,
    "tlm_version" / Byte,
    "voltage" / Int16ub,
    "temperature" / Int16ub,
    "advertising_count" / Int32ub,
    "seconds_since_boot" / Int32ub,
)
frame = GreedyRange(UnencryptedTLMFrame)
parsed = frame.parse(b' \x00\x0b\xa3\x14\x80\x00\xc5\xf9\x03\x10\t\x08!')
parsed[0].voltage /= float(1000)
parsed[0].temperature /= float(256)
```

with struct it is (in the end you only missed the `>`)
```python
frame_type, version, voltage, temperature, count, time = struct.unpack(">BBHHII", b' \x00\x0b\xa3\x14\x80\x00\xc5\xf9\x03\x10\t\x08!')
voltage /= 1000
temperature /= 256
```
